tree/ntuple/v7/src/RNTupleDescriptor.cxx

```diff
@@ -1035,6 +1035,29 @@ void ROOT::Experimental::Internal::RNTupleDescriptorBuilder::BeginHeaderExtensio
       fDescriptor.fHeaderExtension = std::make_unique<RNTupleDescriptor::RHeaderExtension>();
 }
 
+void ROOT::Experimental::Internal::RNTupleDescriptorBuilder::ShiftAliasColumns(std::uint32_t offset)
+{
+   std::vector<DescriptorId_t> aliasColumns;
+   aliasColumns.reserve(fDescriptor.GetNLogicalColumns() - fDescriptor.GetNPhysicalColumns());
+   for (const auto &[_, c] : fDescriptor.fColumnDescriptors) {
+      if (c.IsAliasColumn())
+         aliasColumns.emplace_back(c.GetLogicalId());
+   }
+
+   for (auto id : aliasColumns) {
+      auto c = fDescriptor.fColumnDescriptors[id].Clone();
+      fDescriptor.fColumnDescriptors.erase(id);
+      for (auto &link : fDescriptor.fFieldDescriptors[c.fFieldId].fLogicalColumnIds) {
+         if (link == c.fLogicalColumnId) {
+            link += offset;
+            break;
+         }
+      }
+      c.fLogicalColumnId += offset;
+      fDescriptor.fColumnDescriptors.emplace(c.fLogicalColumnId, std::move(c));
+   }
```

Comment:
Initially this looked more complicated than needed, and especially the nested loop scared me a bit. Now I realize this only iterates over the columns of the projected field that the currently edited column references? So the inner loop has only very few iterations, and not thousands?

Otherwise, if I understand the new invariant correctly, alias columns will always have IDs `fDescriptor.GetNPhysicalColumns()` to `fDescriptor.GetNLogicalColumns()`. Then we should only need one loop over those IDs and adjust them accordingly.

Reply:
> Initially this looked more complicated than needed, and especially the nested loop scared me a bit. Now I realize this only iterates over the columns of the projected field that the currently edited column references? So the inner loop has only very few iterations, and not thousands?

Yes

> Otherwise, if I understand the new invariant correctly, alias columns will always have IDs `fDescriptor.GetNPhysicalColumns()` to `fDescriptor.GetNLogicalColumns()`. Then we should only need one loop over those IDs and adjust them accordingly.

Good point! Done (and uncovered a bug along the way: we need to ensure that the columns are moved from top to bottom. The previous version ended up doing this by chance [unordered map iteration]).